### PR TITLE
Trim silence and cut audio to fixed size

### PR DESCRIPTION
**Pull request by ibab**
I've moved the background audio processing into a separate `utils.py` file and added two features:
 - Cutting the audio samples to a fixed size each time
 - Removing silence from the recordings

It would be great if someone could review the code.

### CODE REVIEW

File: util.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    def get_inputs(self):
+        self.audio_batch, _ = self.queue.dequeue_many(self.args.batch_size)
+        return self.audio_batch, _
+
+    def thread_main(self, sess):
+        buff = np.array([])
+        stop = False
+        # Go through the dataset multiple times
+        while not stop:
+            for audio, label in iterate_through_vctk(self.args.data_dir, self.wavenet_params["sample_rate"]):
```
Comment (lemonzi) [inline util.py]: How about we give CustomRunner an iterable to consume to decouple it form VCTK? It could also have its own file, since it's a separate class.

File: util.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+def iterate_through_directory(directory, sample_rate):
+    files = scan_directory(directory)
+    for f in files:
+        audio, sr = librosa.load(f, sr=sample_rate, mono=True)
+        audio = audio.reshape(-1, 1)
+        yield audio, 0
+
+
+def iterate_through_vctk(directory, sample_rate):
```
Comment (lemonzi) [inline util.py]: Make a folder with dataset wrappers and put this in `vctk.py`?

File: util.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+def scan_directory(directory):
+    files = []
+    for root, dirnames, filenames in os.walk(directory):
+        for filename in fnmatch.filter(filenames, '*.wav'):
+            files.append(os.path.join(root, filename))
+    return files
+
+
+def iterate_through_directory(directory, sample_rate):
```
Comment (lemonzi) [inline util.py]: I would rename this function to something with `audio` in it.

File: util.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+import librosa
+import numpy as np
+import tensorflow as tf
+
+
+def scan_directory(directory):
+    files = []
+    for root, dirnames, filenames in os.walk(directory):
+        for filename in fnmatch.filter(filenames, '*.wav'):
```
Comment (lemonzi) [inline util.py]: `*.wav` as a function argument with a default value?

File: util.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            for audio, label in iterate_through_vctk(self.args.data_dir, self.wavenet_params["sample_rate"]):
+                if self.coord.should_stop():
+                    self.stop_threads()
+                    stop = True
+                    break
+                # Remove silence
+                audio = trim_sample(audio[:,0])
+                # Always cut samples into fixed size pieces
+                buff = np.append(buff, audio)
+                size = 100000
```
Comment (lemonzi) [inline util.py]: This should be given to the constructor of the class.

File: util.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                while len(buff) > size:
+                    piece = np.reshape(buff[:size], [-1, 1])
```
Comment (lemonzi) [inline util.py]: I don't know if we intend to support multi-channel audio at some point. I guess this is fine for now.

File: util.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                    buff = buff[size:]
+
+    def stop_threads():
+        for t in self.threads:
+            t.stop()
+
+    def start_threads(self, sess, n_threads=1):
+        for n in range(n_threads):
+            t = threading.Thread(target=self.thread_main, args=(sess,))
+            t.daemon = True # thread will close when parent quits
```
Comment (lemonzi) [inline util.py]: [readability]: I really like Google's style guide, with 2 spaces before the `#`, first work capitalized and dot at the end. Up to you though.

File: util.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                    piece = np.reshape(buff[:size], [-1, 1])
+                    sess.run(self.enqueue, feed_dict={self.dataX: piece, self.dataY: label})
+                    buff = buff[size:]
+
+    def stop_threads():
+        for t in self.threads:
+            t.stop()
+
+    def start_threads(self, sess, n_threads=1):
+        for n in range(n_threads):
```
Comment (lemonzi) [inline util.py]: `n` is unused, use `_` instead.

File: util.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            t = threading.Thread(target=self.thread_main, args=(sess,))
```
Comment (lemonzi) [inline util.py]: [minor] Rename to `thread`?

File: util.py (hunk 1 above)
Comment (lemonzi) [inline util.py]: Also, this line is too long anyways; in my refactor I had split it by creating the iterator first.

File: util.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+                    stop = True
+                    break
+                # Remove silence
+                audio = trim_sample(audio[:,0])
+                # Always cut samples into fixed size pieces
+                buff = np.append(buff, audio)
+                size = 100000
+                while len(buff) > size:
+                    piece = np.reshape(buff[:size], [-1, 1])
+                    sess.run(self.enqueue, feed_dict={self.dataX: piece, self.dataY: label})
```
Comment (lemonzi) [inline util.py]: Isn't this too long?

```python
sess.run(self.enqueue, feed_dict={self.dataX: piece,
                                  self.dataY: label})
```

File: util.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+class CustomRunner(object):
+    def __init__(self, args, wavenet_params, coord):
+        self.args = args
+        self.wavenet_params = wavenet_params
+        self.coord = coord
+        self.threads = []
+        self.dataX = tf.placeholder(dtype=tf.float32, shape=None)
+        self.dataY = tf.placeholder(dtype=tf.int32)
+        self.queue = tf.PaddingFIFOQueue(
+            256,  # Queue size.
```
Comment (lemonzi) [inline util.py]: Make queue size a parameter.

File: util.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    '''Removes silence in the beginning and end of a sample'''
+    energy = librosa.feature.rmse(audio)
+    indices = librosa.core.frames_to_samples(np.nonzero(energy > threshold))[1]
+    return audio[indices[0]:indices[-1]]
+
+
+class CustomRunner(object):
+    def __init__(self, args, wavenet_params, coord):
+        self.args = args
+        self.wavenet_params = wavenet_params
```
Comment (lemonzi) [inline util.py]: We only need sample_rate from here, no need to pass the whole dictionary.

File: util.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        audio, sr = librosa.load(f, sr=sample_rate, mono=True)
+        audio = audio.reshape(-1, 1)
+        speaker_id, recording_id = [int(i) for i in speaker_re.findall(f)[0]]
+        yield audio, speaker_id
+
+
+def trim_sample(audio, threshold=0.3):
+    '''Removes silence in the beginning and end of a sample'''
+    energy = librosa.feature.rmse(audio)
+    indices = librosa.core.frames_to_samples(np.nonzero(energy > threshold))[1]
```
Comment (lemonzi) [inline util.py]: I would split this in 2 lines for readability.

File: train.py [hunk 14]
```diff
@@ -178,43 +146,6 @@ def validate_directories(args):
     }
 
 
-class CustomRunner(object):
-    def __init__(self, args, wavenet_params, coord):
-        self.args = args
-        self.wavenet_params = wavenet_params
-        self.coord = coord
-        self.threads = []
-        self.dataX = tf.placeholder(dtype=tf.float32, shape=None)
```
Comment (lemonzi) [inline train.py]: The name is too generic. `self.sample_placeholder`?

File: train.py [hunk 15]
```diff
@@ -178,43 +146,6 @@ def validate_directories(args):
...
-        self.dataY = tf.placeholder(dtype=tf.int32)
```
Comment (lemonzi) [inline train.py]: Same here. `self.label_placeholder`?

File: train.py (hunk 14 above)
Comment (lemonzi) [inline train.py]: (they also apply to the new code)

File: audio_reader.py [hunk 16]
```diff
@@ -37,38 +37,40 @@ def load_vctk_audio(directory, sample_rate):
...
     def thread_main(self, sess):
         buff = np.array([])
         stop = False
         # Go through the dataset multiple times
         while not stop:
-            iterator = load_vctk_audio(
-                self.args.data_dir,
-                self.wavenet_params["sample_rate"])
-            for audio, label in iterator:
+            iterator = load_generic_audio( self.audio_dir, self.sample_rate)
```
Comment (lemonzi) [inline audio_reader.py]: Extra space before `self.audio_dir`.

**Comment by ibab** [comment]
Excellent suggestions!
I've switched to using the generic audio loading, as we're not using the VCTK speaker IDs at the moment.
We can do something clever for switching between `AudioReader` and a new `VCTKReader` once we implement conditioning on extra variables.

File: audio_reader.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    for f in files:
+        audio, sr = librosa.load(f, sr=sample_rate, mono=True)
+        audio = audio.reshape(-1, 1)
+        yield audio
+
+
+def load_vctk_audio(directory, sample_rate):
+    files = find_files(directory)
+    speaker_re = re.compile(r'p([0-9]+)_([0-9]+)\.wav')
+    for j, f in enumerate(files):
```
Comment (lemonzi) [inline audio_reader.py]: Why `j` and `enumerate`?

File: audio_reader.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        audio, sr = librosa.load(f, sr=sample_rate, mono=True)
+        audio = audio.reshape(-1, 1)
+        speaker_id, recording_id = [int(i) for i in speaker_re.findall(f)[0]]
```
Comment (lemonzi) [inline audio_reader.py]: s/`i`/`id` ?

File: audio_reader.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        # Go through the dataset multiple times
+        while not stop:
+            iterator = load_generic_audio(self.audio_dir, self.sample_rate)
+            for audio in iterator:
+                if self.coord.should_stop():
+                    self.stop_threads()
+                    stop = True
+                    break
+                # Remove silence
+                audio = trim_sample(audio[:,0])
```
Comment (lemonzi) [inline audio_reader.py]: `[:,0]` -> `[:, 0]`

File: audio_reader.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                                         ['float32'],
+                                         shapes=[(None, 1)])
+        self.enqueue = self.queue.enqueue([self.sample_placeholder])
+
+    def dequeue(self, num_elements):
+        output = self.queue.dequeue_many(num_elements)
+        return output
+
+    def thread_main(self, sess):
+        buff = np.array([])
```
Comment (lemonzi) [inline audio_reader.py]: `buffer`? It might make some lines too long though.

File: audio_reader.py [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+import os
+import re
+import threading
+
+import librosa
+import numpy as np
+import tensorflow as tf
+
+
+def find_files(directory, pattern='*.wav'):
```
Comment (lemonzi) [inline audio_reader.py]: Add a 1-line docstring?

File: audio_reader.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+    files = []
+    for root, dirnames, filenames in os.walk(directory):
+        for filename in fnmatch.filter(filenames, pattern):
+            files.append(os.path.join(root, filename))
+    return files
+
+
+def load_generic_audio(directory, sample_rate):
```
Comment (lemonzi) [inline audio_reader.py]: Add a 1-line docstring?

File: audio_reader.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+    files = find_files(directory)
+    for f in files:
+        audio, sr = librosa.load(f, sr=sample_rate, mono=True)
+        audio = audio.reshape(-1, 1)
+        yield audio
+
+
+def load_vctk_audio(directory, sample_rate):
```
Comment (lemonzi) [inline audio_reader.py]: Add a 1-line docstring?

File: audio_reader.py [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+
+def trim_sample(audio, threshold=0.3):
+    '''Removes silence in the beginning and end of a sample'''
+    energy = librosa.feature.rmse(audio)
+    frames = np.nonzero(energy > threshold)
+    indices = librosa.core.frames_to_samples(frames)[1]
+    return audio[indices[0]:indices[-1]]
+
+
+class AudioReader(object):
```
Comment (lemonzi) [inline audio_reader.py]: Missing docstrings for the class, constructor (args),  and methods. This can go in a separate PR though.

**Comment by lemonzi** [comment]
About isolating VCTK, I mentioned that because a few of us are very interested in music-related applications and we will diverge to other datasets as soon as the network is working. Not a priority for now though, let's teach the system to speak first.

File: audio_reader.py (hunk 18 above)
Comment (ibab) [inline audio_reader.py]: I can use `id_`, as `id` is a keyword.

File: audio_reader.py (hunk 20 above)
Comment (ibab) [inline audio_reader.py]: `buffer` is also a built-in, so I'll use `buffer_`.

**Comment by ibab** [comment]
@lemonzi: Yeah, ideally in the end there should be a nice API for using arbitrary datasets with extra conditioning information.

**Comment by ibab** [comment]
Should I merge now, or can you spot more improvements?

File: audio_reader.py [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+    for root, dirnames, filenames in os.walk(directory):
+        for filename in fnmatch.filter(filenames, pattern):
+            files.append(os.path.join(root, filename))
+    return files
+
+
+def load_generic_audio(directory, sample_rate):
+    '''Generator that yields audio waveforms from the directory.'''
+    files = find_files(directory)
+    for f in files:
```
Comment (lemonzi) [inline audio_reader.py]: s/f/filename ?

File: audio_reader.py [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+        audio, sr = librosa.load(f, sr=sample_rate, mono=True)
```
Comment (lemonzi) [inline audio_reader.py]: `sr` is unused (and redundant), use `_`?

File: audio_reader.py [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+        yield audio
+
+
+def load_vctk_audio(directory, sample_rate):
+    '''Generator that yields audio waveforms from the VCTK dataset, and
+    additionally the ID of the corresponding speaker.'''
+    files = find_files(directory)
+    speaker_re = re.compile(r'p([0-9]+)_([0-9]+)\.wav')
+    for filename in files:
+        audio, sr = librosa.load(filename, sr=sample_rate, mono=True)
```
Comment (lemonzi) [inline audio_reader.py]: `sr` is also unused here.

File: audio_reader.py [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+        audio = audio.reshape(-1, 1)
+        matches = speaker_re.findall(filename)[0]
+        speaker_id, recording_id = [int(id_) for id_ in matches]
+        yield audio, speaker_id
+
+
+def trim_sample(audio, threshold=0.3):
+    '''Removes silence in the beginning and end of a sample.'''
```
Comment (lemonzi) [inline audio_reader.py]: s/in/at

File: audio_reader.py [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+    speaker_re = re.compile(r'p([0-9]+)_([0-9]+)\.wav')
+    for filename in files:
+        audio, sr = librosa.load(filename, sr=sample_rate, mono=True)
+        audio = audio.reshape(-1, 1)
+        matches = speaker_re.findall(filename)[0]
+        speaker_id, recording_id = [int(id_) for id_ in matches]
+        yield audio, speaker_id
+
+
+def trim_sample(audio, threshold=0.3):
```
Comment (lemonzi) [inline audio_reader.py]: I would rename this to something with `silence` in it. As for the method itself, it's very naive but I guess it'll do it. The librosa guys suggest making the threshold relative to the maximum amplitude, but it's a bit dangerous in some situations.

Comment (ibab) [inline audio_reader.py]: That's a good idea and and wouldn't be too difficult to implement.
The current implementation sounded fine when inspecting the concatenated audio, but it can certainly be improved.

Comment (lemonzi) [inline audio_reader.py]: Ok, I'll open an issue to discuss it.

**Comment by lemonzi** [comment]
LGTM, great job with the clean-up! 😄

**Comment by lemonzi** [comment]
Closes #19.

**Comment by ibab** [comment]
You've done a great job reviewing the PR, so I've made you a collaborator 👍 
Feel free to review and merge PRs in the future.

**Comment by lemonzi** [comment]
Wow, this is great, thanks! 😄
